Source/DataSources/EntityCollection.js

```diff
@@ -30,15 +30,27 @@ define([
     };
 
     function fireChangedEvent(collection) {
+        if (collection._firing) {
```

Comment:
Okay, this should be fixed.  I could add a more elaborate test if you think it's worth it that uses multiple spies to make sure things are called in the expected order.